// @dfuse/eosws-js/dist/client/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const outbound_1 = require("../message/outbound");
const listeners_1 = require("./listeners");
class EoswsClient {
    constructor(socket) {
        this.socket = socket;
        this.listeners = new listeners_1.EoswsListeners();
    }
    connect() {
        return this.socket.connect((message) => {
            this.listeners.handleMessage(message);
        });
    }
    disconnect() {
        return this.socket.disconnect();
    }
    getActionTraces(data, options = {}) {
        options = mergeDefaultsStreamOptions(options, {
            listen: true
        });
        return this.createListenerWithSend(outbound_1.getActionTracesMessage(data, options));
    }
    getTableRows(parameters, options = {}) {
        options = mergeDefaultsStreamOptions(options, {
            listen: true
        });
        return this.createListenerWithSend(outbound_1.getTableRowsMessage(parameters, options));
    }
    getTransactionLifecycle(id, options = {}) {
        options = mergeDefaultsStreamOptions(options, {
            fetch: true,
            listen: true
        });
        return this.createListenerWithSend(outbound_1.getTransactionLifecycleMessage({ id }, options));
    }
    createListenerWithSend(message) {
        const reqId = message.req_id;
        const onMessage = (callback) => {
            this.listeners.addListener({ reqId, callback });
            this.socket.send(message);
        };
        return {
            onMessage,
            reqId,
            unlisten: () => this.unlisten(reqId)
        };
    }
    unlisten(requestId) {
        this.listeners.removeListener(requestId);
        this.socket.send(outbound_1.unlistenMessage({ req_id: requestId }));
    }
}
exports.EoswsClient = EoswsClient;
function randomReqId() {
    return `r${Math.random()
        .toString(16)
        .substr(2)}`;
}
function mergeDefaultsStreamOptions(userDefinedOptions, defaultOptions) {
    return Object.assign({ req_id: randomReqId() }, defaultOptions, userDefinedOptions);
}
