// @dfuse/eosws-js/dist/client/listeners.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const debug_1 = __importDefault(require("debug"));
class EoswsListeners {
    constructor(options = {}) {
        this.registeredListeners = [];
        this.debug = debug_1.default("eosws:listeners" + (options.id !== undefined ? `:${options.id}` : ""));
    }
    handleMessage(message) {
        this.registeredListeners.forEach((listener) => {
            if (message.req_id !== listener.reqId) {
                return;
            }
            this.debug("Found matching listener for request id [%s], forwarding message of type [%s].", message.req_id, message.type);
            listener.callback(message);
        });
    }
    addListener(listener) {
        this.debug("Adding a listener for requestId [%s].", listener.reqId);
        this.registeredListeners.push(listener);
    }
    removeListener(reqId) {
        this.debug("Removing listener for requestId [%s] (if it exists).", reqId),
            (this.registeredListeners = this.registeredListeners.filter((listener) => {
                return reqId !== listener.reqId;
            }));
    }
}
exports.EoswsListeners = EoswsListeners;
